freqtrade/exchange/exchange.py

```diff
@@ -1274,6 +1275,8 @@ def create_order(
                 rate_for_order,
                 params,
             )
+            if not self._ft_has.get("create_order_has_all_data"):
+                order = self._api.fetch_order(order["id"], pair)
```

Comment:
ok, the initial issue I noticed was that, when entering an order, freqtrade logged (and sent telegram messages) that stated `amount: 0`. Traced that back to that information not being available in the `create_order()` return info of ccxt. Then saw that the hyperliquid api doesn't return it as well (so it can't be fixed in ccxt). 

How does the kucoin implementation deal with that.. I would expect it to to log / send telegram messages with `amount: 0` ?

I was also worried about timing issues. But, what I found it that hyperliquid always (well, I tested with about 100 positions) returns that info when doing fetch_order directly after creating it. So it looks to be reliable. 

I could move that fetch_order logic to the hyperliquid class in a similar manner to kucoin (see below). Pro's is that it doesn't introduce another config variable. Cons is that once another exchange needs it, that config variable might be cleaner.

```
    def create_order(
        self,
        *,
        pair: str,
        ordertype: str,
        side: BuySell,
        amount: float,
        rate: float,
        leverage: float,
        reduceOnly: bool = False,
        time_in_force: str = "GTC",
    ) -> dict:
        res = super().create_order(
            pair=pair,
            ordertype=ordertype,
            side=side,
            amount=amount,
            rate=rate,
            leverage=leverage,
            reduceOnly=reduceOnly,
            time_in_force=time_in_force,
        )
        # Hyperliquid returns only the order-id.

        if not self._config["dry_run"]:
           res = self._api.fetch_order(res['id'], pair)
        return res
```

let me know what you think!

Reply:
well moving it to hyperliquid won't change the point - if the order is placed _but_ the `fetch_order` call doesn't receive the order yet (timing problem, sometimes seen during peak times) - freqtrade would "forget" this order - as you get an "order not found" error while placing an order - so from a freqtrade perspective - no order was ever placed - hence the trade is never stored (if it happens on the initial order).

---

With kucoin, the result from the exchange contains _only_ the id.
CCXT provides us with that - and all other information as None / undefined.

if hyperliquid returns actual values (`0`, not `None`) - then that should be fixed in ccxt - as that'd mean they're inventing values that are not there. Values not passed by the upstream API and without ability to derrive from other values (like - calculating cost as amount * price) - should be set to None / undefined.

The only thing ccxt does wrong on kucoin is the order status - which we simply override in the response.
If ccxt doesn't fix this - i think the better option is to "null" all values the api doesn't provide (pretty sure they do - as i think that's an undetected bug so far) - which will be everything except for the actual id i'd guess.

https://github.com/freqtrade/freqtrade/blob/1bf71b47df1ccdb8520524d55c6a54f62fd59a6c/freqtrade/exchange/kucoin.py#L65-L68

Reply:
I think ccxt is ok, here's what ccxt returned:

```
2024-10-17-09_10/freqtrade.log:2024-10-16 18:05:13,663 - freqtrade.exchange.exchange - INFO - Order create_order(ETH/USDC:USDC, limit, buy, 0.0196, 2624.9, {'timeInForce': 'GTC'}
2024-10-17-09_10/freqtrade.log:2024-10-16 18:05:14,876 - freqtrade.exchange.exchange - INFO - API create_order: {'info': {'resting': {'oid': '42205402675'}}, 'id': '42205402675', 'clientOrderId': None, 'timestamp': None, 'datetime': None, 'lastTradeTimestamp': None, 'lastUpdateTimestamp': None, 'symbol': None, 'type': 'limit', 'timeInForce': None, 'postOnly': None, 'reduceOnly': None, 'side': None, 'price': None, 'triggerPrice': None, 'amount': None, 'cost': None, 'average': None, 'filled': None, 'remaining': None, 'status': 'open', 'fee': None, 'trades': [], 'fees': [], 'stopPrice': None, 'takeProfitPrice': None, 'stopLossPrice': None}
```

so amount is `None`.. Which for some reason seemed to result in the telegram message amount being 0 (because after doing the explicit fetch_order, I did not see that anymore)

```
2024-10-17-09_10/freqtrade.log:2024-10-16 18:05:15,795 - freqtrade.rpcrpc.rpc_manager - INFO - Sending rpc message: {'trade_id': 6, 'type': entry, 'buy_tag': 'timestamp_enter_long', 'enter_tag': 'timestamp_enter_long', 'exchange': 'Hyperliquid', 'pair': 'ETH/USDC:USDC', 'leverage': 2.0, 'direction': 'Long', 'limit': 2624.9, 'open_rate': 2624.9, 'order_type': 'limit', 'stake_amount': 25.790319685350003, 'stake_currency': 'USDC', 'base_currency': 'ETH', 'quote_currency': 'USDC', 'fiat_currency': 'USD', 'amount': 0.0, 'open_date': datetime.datetime(2024, 10, 16, 18, 5, 14, 879725, tzinfo=datetime.timezone.utc), 'current_rate': 2624.9, 'sub_trade': False}
```

Now looking at the code I don't really understand why that happened.. need to investigate it again

Reply:
ok, so in freqtradebot.py `execute_entry()` 

* order['amount'] is `None`. 
* [here](https://github.com/freqtrade/freqtrade/blob/63dfa52ea85c2d8f9dcabba5c457f0dc7c75e0b4/freqtrade/freqtradebot.py#L1002) trade.amount is set to 0. 

in `_notify_enter()` [here](https://github.com/freqtrade/freqtrade/blob/63dfa52ea85c2d8f9dcabba5c457f0dc7c75e0b4/freqtrade/freqtradebot.py#L1216C13-L1216C95) it says `"amount": order.safe_amount_after_fee if fill else (order.amount or trade.amount),`

I logged that: `2024-11-03 22:05:15,452 - freqtrade.freqtradebot - INFO - order.safe_amount_after_fee: 0.0 fill: False order.amount: None trade.amount: 0.0`

So.. that explains why the telegram message is sent with amount 0. Not sure how it works for kucoin?

Reply:
it works the same way - and has just been fixed in develop (in a completely different way).

I don't see a reason to call `fetch_order` or do any other shenanigans - as freqtrade handles a response from create_order like `{'id': 'orderidasdf123'}` perfectly fine (proven by kucoin, actually).

Reply:
ok, https://github.com/freqtrade/freqtrade/commit/5cb4a65cbd67b36e8b10ef34d999d62f2ad9ac8a I guess. If it's fixed there I can of course delete my workarounds. I'll test it.

Reply:
seems to work! /status shows Amount: 0 for a few seconds though, not sure if that's something you want to fix. I'll revert my 'workaround' commits.

Reply:
done https://github.com/freqtrade/freqtrade/pull/10839/commits/38eb4eed97bfc408e6324503d02e98d65c0f359e

Reply:
> seems to work! /status shows Amount: 0 for a few seconds though, not sure if that's something you want to fix. I'll revert my 'workaround' commits.

no that's an intentional change. 
You don't actually own anything until the order actually fills (think - limit order that never fills) - so the amount that was shown previously was wrong.